### PR TITLE
Simulated Engine Implementation

### PR DESCRIPTION
**Pull request by dstrain115**
Add AbstractEngine Interface
    
    This PR adds three groups of classes to cirq_google.
    
    The first four classes define abstract base classes (ABC)
    for the four main types of the Engine:
    
    AbstractJob, AbstractProgram, AbstractProcessor, AbstractEngine
    
    These classes closely model the currect EngineJob, EngineProgram,
    EngineProcessor, and Engine classes with the methods turned into
    abstract models.
    
    The second group is a partially implemented sub-classes that define
    in-memory objects for functionality unlikely to be implemented
    differently, such as reservations, labels, and descriptions:
    
    AbstractLocalJob, AbstractLocalProgram,
    AbstractLocalProcessor, AbstractLocalEngine
    
    Lastly, the final group of classes defines implementation for
    an Engine emulator backed by a sampler, such as the cirq simulator
    or other local simulator.
    
    SimulatedLocalJob, SimulatedLocalProgram,
    SimulatedLocalProcessor, SimulatedLocalEngine

### CODE REVIEW

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
Can we break this up into smaller pieces ?

**Comment by dstrain115** [comment]
I merged with master, which brought file count down from 25 to 20.

I have split this into 3 PRs:  
1) Abstract interface only, no functionality (4 files) #4644
2) Partial local implementation (8 additional files) #4645
3) Full implementation using simulated sampler (8 additional files) #4638 (This PR)

Once the first two PRs are merged, this will be down to 8 files.

**Comment by dstrain115** [comment]
Ok, the first two PRs are in and merged.  This PR is ready to review with only 8 files now!

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:69]: `sammpler` -> `sampler`

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:57]: `VAlidation` -> `Validation`

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:114]: [Optional] There's probably a more concise/pythonic way of doing this with `max(..., key=...)`. Leaving it up to you whether to switch to that form.

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:129]: Add negative test for this method.

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:319]: Add negative test

File: cirq-google/cirq_google/engine/simulated_local_program.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_program.py:25]: Tests for this class? Kind of surprised coverage check isn't firing for this...

File: cirq-google/cirq_google/engine/simulated_local_program.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_program.py:31]: Remove this line? This doesn't look like an abstract class, at least.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
First pass - I still need to do a deeper dive on `simulated_local_processor[_test].py`.

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (MichaelBroughton) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:178]: Nit: can probably just use list comprehension here too.

**Review by MichaelBroughton — COMMENTED** [review]
Just a few comments from me and then I think we should be good to merge.

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:36]: Since `run_calibration` isn't implemented, should we remove `FocusedCalibration` from this list?

File: cirq-google/cirq_google/engine/simulated_local_processor_test.py
Comment (95-martin-orion) [inline cirq-google/cirq_google/engine/simulated_local_processor_test.py:161]: Missing test block? Otherwise remove comment.

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Finished `simulated_local_processor_test.py` review, just a few additional items.

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/simulated_local_processor.py]: Done.

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/simulated_local_processor.py]: Done.

File: cirq-google/cirq_google/engine/simulated_local_engine_test.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/simulated_local_engine_test.py:30]: I am loathe to remove these tests.  Even if they are redundant, it is nice to test that all the cross-references work with the final child class.

File: cirq-google/cirq_google/engine/simulated_local_job.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/simulated_local_job.py:113]: I am not sure I understand the question.  This line uses run() which returns a single Result.  The line above uses run_sweep which returns a List[Result].

File: cirq-google/cirq_google/engine/simulated_local_program.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/simulated_local_program.py:25]: Added some basic tests.  Coverage was previously provided by tests in processor and engine.

File: cirq-google/cirq_google/engine/simulated_local_job.py
Comment (MichaelBroughton) [reply cirq-google/cirq_google/engine/simulated_local_job.py:113]: Sorry, let me ask instead: Why couldn't we do something like:
```python3
self._state = quantum.enums.thing
return self._sampler.run_sweep(program=program, sweeps[0] if sweeps else None, repetitions=reps)
```
and get rid of the need for the if/else block ?

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (MichaelBroughton) [inline cirq-google/cirq_google/engine/simulated_local_processor.py:167]: Perhaps storing an intermediate list or two here to break up the big compound expression might be a little easier to parse as a reader ?

**Review by MichaelBroughton — APPROVED** [review]
LGTM, left two final comments, feel free to ignore if you want.

File: cirq-google/cirq_google/engine/simulated_local_program.py
Comment (95-martin-orion) [reply cirq-google/cirq_google/engine/simulated_local_program.py:25]: Where are these tests? There's no `simulated_local_program_test.py` in this PR - did it get missed in the commit?

File: cirq-google/cirq_google/engine/simulated_local_engine_test.py
Comment (95-martin-orion) [reply cirq-google/cirq_google/engine/simulated_local_engine_test.py:30]: SGTM.

**Review by 95-martin-orion — APPROVED** [review]
LGTM once `simulated_local_program_test` comment is resolved.

File: cirq-google/cirq_google/engine/simulated_local_processor.py
Comment (dstrain115) [reply cirq-google/cirq_google/engine/simulated_local_processor.py:167]: Made an intermediate function instead.

**Pull request merged**
